Unified Web API and 2.2 version (#9469)
aspnet/Docs: aspnetcore/tutorials/first-web-api.md

---

> title: Create a web api with ASP.NET Core and Visual Studio

Reviewer: Make the following change all throughout:

api --> API

author: rick-anderson
description: Build a web api with ASP.NET Core MVC and Visual Studio
ms.author: riande
monikerRange: '>= aspnetcore-2.1'
ms.custom: mvc
ms.date: 11/19/2018
uid: tutorials/first-web-api
---
# Tutorial: Create a web api with ASP.NET Core

By [Rick Anderson](https://twitter.com/RickAndMSFT) and [Mike Wasson](https://github.com/mikewasson)

> This tutorial teaches the basics of building a web api app. The app manages a list of "to-do" items. A user interface (UI) isn't created. You learn how to:

Reviewer: ```suggestion
This tutorial teaches the basics of building a web API to manage a list of "to-do" items. A user interface (UI) isn't created. You learn how to:
```

> [!div class="checklist"]
> * Create a web api project.
> * Add a controller.
> * Add code to get "to-do" items.
> * Create other CRUD operations.
> * Call the web api with jQuery.

> At the end, you have an app that can manage "to-do" items.

Reviewer: ```suggestion
At the end, you have a web API that can manage "to-do" items.
```

Right-click the *Models* folder, and select **Add** > **New File** > **General** > **Empty Class**. Name the class *TodoItem*, and then click **New**.

Update the `TodoItem` class with the following code:

---

[!code-csharp[](first-web-api/samples/2.2/TodoApi/Models/TodoItem.cs)]

Model classes can go anywhere in the project, but the *Models* folder is used by convention.

The database generates the `Id` when a `TodoItem` is created.

### Create the database context

The *database context* is the main class that coordinates Entity Framework functionality for a given data model. This class is created by deriving from the `Microsoft.EntityFrameworkCore.DbContext` class.

# [Visual Studio](#tab/visual-studio)

Right-click the *Models* folder and select **Add** > **Class**. Name the class *TodoContext* and click **Add**.

# [Visual Studio Code](#tab/visual-studio-code)

Add a `TodoContext` class to the *Models* folder.

# [Visual Studio for Mac](#tab/visual-studio-mac)

Add a `TodoContext` class in the *Models* folder:

---

Replace the class with the following code:

[!code-csharp[](first-web-api/samples/2.2/TodoApi/Models/TodoContext.cs)]

## Register the database context

> In this step, the database context is registered with the [dependency injection](xref:fundamentals/dependency-injection) container. Services (such as the DB context) that are registered with the dependency injection (DI) container are available to the controllers.

Reviewer: DB context --> database context

> Register the DB context with the service container using the built-in support for [dependency injection](xref:fundamentals/dependency-injection). Update *Startup.cs* with the following highlighted code:

Reviewer: ```suggestion
Register the DB context with the service container using the built-in support for DI. Update *Startup.cs* with the following highlighted code:
```

::: moniker range="= aspnetcore-2.2"
[!code-csharp[](first-web-api/samples/2.2/TodoApi/Startup1.cs?highlight=5,8,25-26&name=snippet_all)]
::: moniker-end
::: moniker range="= aspnetcore-2.1"
[!code-csharp[](first-web-api/samples/2.1/TodoApi/Startup.cs?highlight=7-8,25-26)]

::: moniker-end

The preceding code:

* Removes unused `using` declarations.
* Specifies an in-memory database is injected into the service container.

### Add a controller

# [Visual Studio](#tab/visual-studio)

* Right-click the *Controllers* folder.
* Select **Add** > **New Item**.
* In the **Add New Item** dialog, select the **API Controller Class** template.
* Name the class *TodoController*, and click **Add**.

![Add new Item dialog with controller in search box and web API controller selected](first-web-api/_static/new_controller.png)

# [Visual Studio Code](#tab/visual-studio-code)

In the *Controllers* folder, create a class named `TodoController`.

# [Visual Studio for Mac](#tab/visual-studio-mac)

> In Solution Explorer, in the *Controllers* folder, add the class `TodoController`.

Reviewer: I've been bolding Solution Explorer, since it's a UI element.

---

Replace the `TodoController` class with the following code:

[!code-csharp[](first-web-api/samples/2.2/TodoApi/Controllers/TodoController2.cs?name=snippet_todo1)]

The preceding code:

* Defines an API controller class without methods.
* Creates a new Todo item when `TodoItems` is empty. You won't be able to delete all the Todo items because the constructor creates a new one if `TodoItems` is empty.

> In the next sections, methods are added to implement the API. The class is annotated with an [`[ApiController]`](/dotnet/api/microsoft.aspnetcore.mvc.apicontrollerattribute) attribute to:

Reviewer: Remove code-fencing, since this is a link.

Author: I'd like to keep in the inner `[]` and `[\[` doesn't work. We can probably make an exception for attributes.

Reviewer 2: Do you really need brackets when you're explicitly identifying it as an attribute?

* Mark the class is used to serve HTTP API responses.
* Enable the class to target conventions, filters, and other behaviors.

> For information, see [Annotation with ApiControllerAttribute](xref:web-api/index#annotation-with-apicontrollerattribute).

Reviewer: ```suggestion
For information, see [Annotation with ApiController attribute](xref:web-api/index#annotation-with-apicontroller-attribute).
```

> The controller's constructor uses [Dependency Injection](xref:fundamentals/dependency-injection) to inject the database context (`TodoContext`) into the controller. The database context is used in each of the [CRUD](https://wikipedia.org/wiki/Create,_read,_update_and_delete) methods in the controller. The constructor adds an item to the in-memory database if one doesn't exist.

Reviewer: ```suggestion
The controller's constructor uses DI to inject the database context (`TodoContext`) into the controller. The database context is used in each of the [CRUD](https://wikipedia.org/wiki/Create,_read,_update_and_delete) methods in the controller. The constructor adds an item to the in-memory database if one doesn't exist.
```

## Get to-do items

To get to-do items, add the following methods to the `TodoController` class:

[!code-csharp[](first-web-api/samples/2.2/TodoApi/Controllers/TodoController.cs?name=snippet_GetAll)]

These methods implement the two GET endpoints:

* `GET /api/todo`
* `GET /api/todo/{id}`

Test the app by calling the two endpoints from a browser. For example:

* `https://localhost:5001/api/todo`
* `https://localhost:5001/api/todo/1`

The following HTTP response is produced with the preceding call to `GetAll`:

```json
[
  {
    "id": 1,
    "name": "Item1",
    "isComplete": false
  }
]
```

Later in the tutorial, instructions are provided to view the HTTP response with [Postman](https://www.getpostman.com/) or [curl](https://curl.haxx.se/docs/manpage.html).

### Routing and URL paths

> The [`[HttpGet]`](/dotnet/api/microsoft.aspnetcore.mvc.httpgetattribute) attribute denotes a method that responds to an HTTP GET request. The URL path for each method is constructed as follows:

Reviewer: Remove code-fencing, since this is a link.
